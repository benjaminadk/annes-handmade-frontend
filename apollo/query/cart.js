import gql from 'graphql-tag'

export const CART_QUERY = gql`
  query {
    cartOpen @client
  }
`
