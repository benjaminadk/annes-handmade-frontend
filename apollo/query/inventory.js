import gql from 'graphql-tag'

export const INVENTORY_QUERY = gql`
  query INVENTORY_QUERY {
    products {
      id
      title
      description
      variant
      bead
      images
      price
      sold
      createdAt
    }
  }
`
