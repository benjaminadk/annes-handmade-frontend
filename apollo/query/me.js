import gql from 'graphql-tag'

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      image
      role
      cart {
        id
        product {
          id
          title
          images
          price
          variant
          bead
        }
      }
      orders {
        id
        charge
        total
        shipped
        createdAt
        items {
          id
          images
        }
      }
    }
  }
`
