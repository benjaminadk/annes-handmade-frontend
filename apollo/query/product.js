import gql from 'graphql-tag'

export const SINGLE_PRODUCT_QUERY = gql`
  query SINGLE_PRODUCT_QUERY($id: ID!) {
    product(id: $id) {
      id
      title
      description
      variant
      bead
      images
      price
    }
    thumbIndex @client
  }
`
