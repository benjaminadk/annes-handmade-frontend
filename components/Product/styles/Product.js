import styled from 'styled-components'
import media from '../../styles/media'

export const ProductStyles = styled.div`
  display: grid;
  grid-template-columns: auto auto 1fr;
  grid-gap: 50px;
  margin-top: 5rem;
  padding: 2rem;
  ${media.tablet`
    grid-template-columns: auto; 
    grid-template-rows: auto auto; 
    grid-gap: 10px;
    justify-items: center;
    margin-top: 1rem;
  `}
`
