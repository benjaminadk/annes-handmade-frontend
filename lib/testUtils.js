import casual from 'casual'
import { mount } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '../components/Page/theme'

const mountWithTheme = component => mount(<ThemeProvider theme={theme}>{component}</ThemeProvider>)

const createEvent = (name, value) => ({ target: { name, value } })

casual.seed(666)

const fakeProduct = () => ({
  __typename: 'Product',
  id: '123',
  title: 'A Fake Product',
  description: 'This is not really real.',
  variant: 'NECKLACE',
  bead: 'RED_JASPER',
  images: ['image1.png', 'image2.png'],
  price: 1000
})

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  role: 'ADMIN',
  image: 'https://fake.png',
  orders: [],
  cart: []
})

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  images: [`${casual.word}.jpg`],
  title: casual.words(),
  price: 4234,
  description: casual.words(),
  variant: 'FAKELACE',
  bead: 'FAKILLIUM'
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  shipped: false,
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser()
})

const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: 'omg123',
  item: fakeProduct(),
  user: fakeUser(),
  ...overrides
})

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key) {
    return this.store[key] || null
  }

  setItem(key, value) {
    this.store[key] = value.toString()
  }

  removeItem(key) {
    delete this.store[key]
  }
}

export {
  LocalStorageMock,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  mountWithTheme,
  createEvent
}
