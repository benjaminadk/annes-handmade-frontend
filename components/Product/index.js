import { Query } from 'react-apollo'
import NextSeo from 'next-seo'
import { SINGLE_PRODUCT_QUERY } from '../../apollo/query/product'
import ProductDetail from './ProductDetail'
import Thumbnails from './Thumbnails'
import LargeImage from './LargeImage'
import Loading from '../Loading'
import DisplayError from '../DisplayError'
import { ProductStyles } from './styles/Product'

export default function Product({ query: { id }, user }) {
  return (
    <Query query={SINGLE_PRODUCT_QUERY} variables={{ id }}>
      {({ data, loading, error }) => {
        if (loading) {
          return <Loading size={10} message="Loading Product" />
        }
        if (error) {
          return <DisplayError error={error} />
        }
        const { product, thumbIndex } = data
        return (
          <>
            <NextSeo
              config={{
                title: product.title,
                canonical: `https://anneshandmade.herokuapp.com/product?id=${id}`,
                openGraph: {
                  title: `Anne's Handmade | ${product.title}`,
                  url: `https://anneshandmade.herokuapp.com/product?id=${id}`,
                  images: [
                    {
                      url: product.images[0],
                      width: 800,
                      height: 800,
                      alt: product.title
                    }
                  ]
                }
              }}
            />
            <ProductStyles>
              <Thumbnails images={product.images} thumbIndex={thumbIndex} />
              <LargeImage image={product.images[thumbIndex]} />
              <ProductDetail product={product} user={user} />
            </ProductStyles>
          </>
        )
      }}
    </Query>
  )
}
