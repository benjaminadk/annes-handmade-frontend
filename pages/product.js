import Product from '../components/Product'

export default props => <Product {...props} />
