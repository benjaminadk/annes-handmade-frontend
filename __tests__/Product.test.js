import { MockedProvider } from 'react-apollo/test-utils'
import wait from 'waait'
import Product from '../components/Product'
import { SINGLE_PRODUCT_QUERY } from '../apollo/query/product'
import { fakeProduct, mountWithTheme } from '../lib/testUtils'

describe('<Product/>', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        request: { query: SINGLE_PRODUCT_QUERY, variables: { id: '123' } },
        result: {
          data: {
            product: fakeProduct(),
            thumbIndex: 0
          }
        }
      }
    ]
    const wrapper = mountWithTheme(
      <MockedProvider mocks={mocks}>
        <Product query={{ id: '123' }} />
      </MockedProvider>
    )
    expect(wrapper.text()).toContain('Loading')
    await wait()
    wrapper.update()
    expect(wrapper.find('p.price')).toMatchSnapshot()
    expect(wrapper.find('p.description')).toMatchSnapshot()
    expect(wrapper.find('p.title')).toMatchSnapshot()
  })

  it('displays error when product is not found', async () => {
    const mocks = [
      {
        request: { query: SINGLE_PRODUCT_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Product not found' }]
        }
      }
    ]
    const wrapper = mountWithTheme(
      <MockedProvider mocks={mocks}>
        <Product query={{ id: '123' }} />
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    const item = wrapper.find('[data-test="graphql-error"]')
    expect(item.text()).toContain('Product not found')
    expect(item).toMatchSnapshot()
  })
})
