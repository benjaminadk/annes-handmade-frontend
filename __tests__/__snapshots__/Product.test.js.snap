// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> displays error when product is not found 1`] = `
<p
  data-test="graphql-error"
>
  Error: 
  Product not found
</p>
`;

exports[`<Product/> renders with proper data 1`] = `null`;

exports[`<Product/> renders with proper data 2`] = `null`;

exports[`<Product/> renders with proper data 3`] = `null`;
